text/0076-entity-slice-validation.md

```diff
@@ -0,0 +1,440 @@
+
+## Alternatives
+
+### Alternative: Entity Slicing using Data Tries
+
+[RFC 74, Entity Slicing using Data Tries](https://github.com/cedar-policy/rfcs/pull/74) proposes the idea of an _entity manifest_ to ease the task of entity slicing. This manifest is a trie-based data structure that expresses which entity attributes and ancestors are needed for each possible action. This data structure is strictly more precise than the concept of level, meaning that it may require less entity data to be provided. For example, the first part of our Basic Example in this RFC (policies 1, 2, and 4 of TinyTodo), the determined entity manifest would indicate that for the `GetList` action, you would need to provide `resource.owner`, `resource.readers`, `resource.editors` and the ancestors of `principal`. This is less data than would be retrieved by the slicing algorithm proposed here: These are level-1 valid policies, so the generic algorithm would acquire the same data given in the manifest but also more, e.g., the ancestors of `resource` and the attributes of `principal`.
+
+Entity manifests are less effective as a _prescriptive_ mechanism that aims to limit the shape of allowable policies. For example, a Cedar policy and entity storage service might like to disallow uploading policies whose entity data is expensive to retrieve in the worst case. With levels, such a service can upper-bound this work by blocking schemas with a level greater than a threshold, say 3. Validation will then ensure that uploaded policies respect this upper bound. Manifests are unwieldy as a specification mechanism, since they are specialized to particular entity definitions and name particular types and attributes, of which there could be many. They'd also have to be updated as types evolved over time, while a level-based specification is less likely to change.
+
+We believe that ultimately level-based validation and entity manifests should coexist. Level-based validation is used to bound entity retrieval work and prevent pathological policies, while entity manifests are used to more efficiently define the needed entity data. We could imagine a deployment strategy in which we accept this RFC and perform generic entity slicing as described here, and then later implement entity manifests, and start performing slicing using those instead.
+
+### Alternative: Level as a validation parameter, not in the schema
+
+This RFC has suggested that the level should be specified in the schema. Alternatively, we could specify the level as a parameter to the validator itself, leaving schemas unchanged. The benefit of doing so is that entity slicing becomes an orthogonal concern: The schema specifies the structure of data, but nothing about what data is required/allowed for a request. The drawback is that policy writers cannot look in one place to know the limits on the policies they can write.  
```

Comment:
I'd argue that the schema _does_ specify what data is required/allowed for a request since it specifies each action's expected principal/resource/context. So I think it makes sense to have level information in the schema too.

Reply:
+1